src/clusterfuzz/_internal/system/archive.py

```diff
@@ -34,6 +33,197 @@
+    self._filename = filename
+    self._is_dir = is_dir
+    self._file_size = file_size
+    self._compress_size = compress_size
+
+  @property
+  def filename(self):
+    return self._filename
+
+  def is_dir(self):
+    raise self._is_dir
+
+  @property
+  def file_size(self):
+    return self._file_size
+
+  @property
+  def compress_size(self):
+    return self._compress_size
+
+
+class ArchiveReader:
+  """ArchiveReader"""
+
+  # This should have a few of the handful functions a tool like unzip has
+  def list_files(self) -> [ArchiveMemberInfo]:
+    raise NotImplementedError
+
+  def extract(self, member, path=None, trusted=False):
+    raise NotImplementedError
+
+  def open(self, member):
+    raise NotImplementedError
+
+  def extractall(self, path=None, members=None, trusted=False) -> None:
+    raise NotImplementedError
+
+  def extracted_size(self, file_match_callback=None):
+    return sum(f.file_size
+               for f in self.list_files()
+               if not file_match_callback or file_match_callback(f.filename))
+
+
+class TarArchiveReader(ArchiveReader):
+  """TarArchiveReader"""
+
+  def __init__(self, archive_path, file_obj=None) -> None:
+    # we need to know whether it's a lzma file or not.
+    archive_type = get_archive_type(archive_path)
+    if file_obj:
+      self.archive = tarfile.open(fileobj=file_obj)
+    else:
+      mode = 'r' if archive_type == ArchiveType.TAR else 'r:xz'
+      self.archive = tarfile.open(archive_path, mode=mode)
+    self.archive_path = archive_path
+
+  def list_files(self) -> [ArchiveMemberInfo]:
+    return [
+        ArchiveMemberInfo(
+            filename=f.name,
+            is_dir=f.isdir(),
+            file_size=f.size,
+            compress_size=f.size) for f in self.archive.getmembers()
+    ]
+
+  def open(self, member):
+    return self.archive.extractfile(member)
+
+  def extract(self, member, path=None, trusted=False):
+    # If the output directory is a symlink, get its actual path since we will be
+    # doing directory traversal checks later when unpacking the archive.
+    output_directory = os.path.realpath(path)
+
+    if not trusted:
+      if (_is_attempting_path_traversal(self.archive_path, output_directory,
+                                        member)):
+        return None
+
+    self.archive.extract(member=member, path=output_directory)
+    return os.path.realpath(os.path.join(output_directory, member))
+
+  def extractall(self, path=None, members=None, trusted=False) -> None:
+    to_extract = members if members is not None else self.archive.namelist()
+    #FIXME(paulsemel): we could try extracting that all.
+    for member in to_extract:
+      self.extract(member=member, path=path, trusted=trusted)
+
+
+class ZipArchiveReader(ArchiveReader):
+  """ZipArchiveReader"""
+
+  def __init__(self, file) -> None:
+    self.zip_archive = zipfile.ZipFile(file, mode='r')
+
+  def list_files(self):
```

Comment:
Consider yielding entries instead, since many usecases below don't need the whole list at once.